### PR TITLE
Upgrade libgit2 to v0.27.2

### PR DESCRIPTION
**Pull request by staticfloat**
This is a possibly-unbroken upgrade path to a newer mbedTLS.  It includes a pull request [I just opened against libgit2](https://github.com/libgit2/libgit2/pull/4678), and in local testing successfully suppresses linking against `libssl.so`.

### CODE REVIEW

**Comment by staticfloat** [comment]
This may be the first PR I have authored in my life to pass only Windows CI and none of the others.

**Comment by simonbyrne** [comment]
I think my PR was only for libgit2 v0.27.1, but v0.27.2 had now been released, so we should use that.

**Comment by staticfloat** [comment]
> I think my PR was only for libgit2 v0.27.1, but v0.27.2 had now been released, so we should use that.

Yep, that's what this is based off of.  I've also switched over to the `CollisionDetection` SHA1 backend as suggested to me in my PR over on the libgit2 repo, as there's no reason not to.

**Comment by simonbyrne** [comment]
the libgit2 changes don't appear to be on this branch (or github is doing something funny with the diff).

**Comment by staticfloat** [comment]
Wow, thanks for that.  A rebase went bad I think.  I `git reflog`'ed and rescued out my commits.

**Comment by simonbyrne** [comment]
You will probably need most of the other changes from #27241.

File: deps/patches/libgit2-mbedtls2.patch [hunk 1]
```diff
@@ -0,0 +1,39 @@
+diff --git a/.travis.yml b/.travis.yml
```
Comment (simonbyrne) [inline deps/patches/libgit2-mbedtls2.patch]: Add a link to the upstream PR?

Comment (staticfloat) [reply deps/patches/libgit2-mbedtls2.patch]: done

File: deps/checksums/libgit2-b0d9952c318a3d1b8917e06ad46b9110c0c28831.tar.gz/md5 [hunk 2]
```diff
@@ -0,0 +1 @@
+6059f607530c302aa1b74c77fcbb5fa2
```
Comment (simonbyrne) [inline deps/checksums/libgit2-b0d9952c318a3d1b8917e06ad46b9110c0c28831.tar.gz/md5]: can remove this checksum (it was for 0.27.1)

File: deps/checksums/libgit2-b0d9952c318a3d1b8917e06ad46b9110c0c28831.tar.gz/sha512 [hunk 3]
```diff
@@ -0,0 +1 @@
+6827d19be048be94d87d118c04c37ee9fc8161d1fb5820fd0feadc8b9d99f08835e1c5ccab47ae2406cb4601efce0d30d03bd66d1ec6e25d591c3ba1e0073ec7
```
Comment (simonbyrne) [inline deps/checksums/libgit2-b0d9952c318a3d1b8917e06ad46b9110c0c28831.tar.gz/sha512]: and this one too

**Comment by simonbyrne** [comment]
LGTM. Will need to check that ssh clones of private repos work on Windows (see https://github.com/JuliaLang/julia/pull/27241#discussion_r193920174)

**Comment by simonbyrne** [comment]
Is there a way we can run the buildbots without merging? Or just try it and see?

**Comment by simonbyrne** [comment]
The 64-bit linux failure seems to persist, and has something to do with this line:
https://github.com/JuliaLang/julia/blob/4420717206c87551ad680fe2ae7add692162a0ff/stdlib/LibGit2/test/libgit2.jl#L2368

**Comment by staticfloat** [comment]
> Is there a way we can run the buildbots without merging? Or just try it and see?

We can, and we should.  We just go to `build.julialang.org`, navigate to the `package_linux64` builder, for instance, click the big blue "package" button, and paste the gitsha of the tip of this branch in.

**Comment by staticfloat** [comment]
If I remember correctly, that credentials thing is part of what will be fixed by https://github.com/JuliaLang/julia/pull/24738.  Essentially, right now a piece of memory is getting overwritten with zeros in a way that can be a little race-conditiony.  I've restarted the build one more time just to make sure.

**Comment by simonbyrne** [comment]
Let's hope so. @mbauman is working on it as we speak.

**Comment by staticfloat** [comment]
Great, I see that you've logged in and [triggered the build](https://build.julialang.org/#/buildrequests/19847).  It will run through everything (including all tests) but we should download it locally and ensure it runs (and doesn't have a dependency on `libssl.so`) before merging this.

**Comment by ViralBShah** [comment]
CI straight flush!

**Comment by simonbyrne** [comment]
Linux looks good:
```
├ ldd julia-a4797b3bb0/lib/julia/libgit2.so
	linux-vdso.so.1 =>  (0x00007ffe1b7f9000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f94c12d8000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f94c10b9000)
	libcurl.so.4 => /home/simonbyrne/julia-a4797b3bb0/lib/julia/libcurl.so.4 (0x00007f94c0e4a000)
	libmbedtls.so.10 => /home/simonbyrne/julia-a4797b3bb0/lib/julia/libmbedtls.so.10 (0x00007f94c0c22000)
	libmbedx509.so.0 => /home/simonbyrne/julia-a4797b3bb0/lib/julia/libmbedx509.so.0 (0x00007f94c0a0e000)
	libmbedcrypto.so.0 => /home/simonbyrne/julia-a4797b3bb0/lib/julia/libmbedcrypto.so.0 (0x00007f94c07bc000)
	libssh2.so.1 => /home/simonbyrne/julia-a4797b3bb0/lib/julia/libssh2.so.1 (0x00007f94c058c000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f94c01ac000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f94c180b000)
```
though it failed the same test as travis.

**Comment by KristofferC** [comment]
Does https://github.com/JuliaLang/julia/pull/27386#issuecomment-396854968 have anything to do with this? This was merged very recently though...

**Comment by staticfloat** [comment]
No, I don't think it could be.  This PR doesn't touch `libcurl` code.

**Comment by simonbyrne** [comment]
The Windows build didn't seem to work, and we don't have new nightlies, so we can't yet test whether ssh works.

**Comment by staticfloat** [comment]
I'm giving some TLC to the windows buildbots right now.  Filesystem permissions problems, as usual.

**Comment by simonbyrne** [comment]
FYI: I tried out ssh clone of a private repo on windows and it seemed to work.

**Comment by nalimilan** [comment]
I get a `ReadOnlyMemoryError` when building RPM nightlies since this PR (I've bisected it): https://copr-be.cloud.fedoraproject.org/results/nalimilan/julia-nightlies/fedora-28-x86_64/00766125-julia/build.log.gz

Ideas?

**Comment by staticfloat** [comment]
Grrr, looks like this also broke `ssh` on OSX.  During `cmake` time, we get:

```
-- Checking for module 'libssh2'
--   Package 'libssl', required by 'libssh2', not found
...
-- Enabled features:
 * threadsafe, threadsafe support
 * cURL, cURL for HTTP proxy support
 * HTTPS, using SecureTransport
 * SHA, using CollisionDetection
 * http-parser, http-parser support
 * zlib, using bundled zlib
 * iconv, iconv encoding conversion support

-- Disabled features:
 * debugpool, debug pool allocator
 * tracing, tracing support
 * SSH, SSH transport support
 * SPNEGO, SPNEGO authentication support
```

This causes [this test](https://github.com/JuliaLang/julia/blob/228e7c65725657c45d3c826562bcc8f0cab8da1d/stdlib/LibGit2/test/libgit2.jl#L156) to fail.

**Comment by simonbyrne** [comment]
Based on the error message, it means that `FIND_PKGLIBRARIES(LIBSSH2 libssh2)` is failing somewhere?

I don't understand CMake well enough though.

**Comment by staticfloat** [comment]
Fixed in https://github.com/JuliaLang/julia/pull/27601

**Comment by staticfloat** [comment]
@nalimilan do you have any extra patches applied to libgit2, libssh2 or libmbedtls?  It looks like we're dying within mbedTLS while initializing mbedTLS while libssh2 is initializing on behalf of libgit2's initialization routines.

**Comment by staticfloat** [comment]
Also, upstream libgit2 patch has landed, huzzah.  Hopefully eventually we can drop these patches.

**Comment by nalimilan** [comment]
> @nalimilan do you have any extra patches applied to libgit2, libssh2 or libmbedtls? It looks like we're dying within mbedTLS while initializing mbedTLS while libssh2 is initializing on behalf of libgit2's initialization routines.

Not AFAICT. I'm using `USE_SYSTEM_LIBGIT2=0 USE_SYSTEM_LIBSSH2=0 USE_SYSTEM_MBEDTLS=0`. Could it be related to the zeroed strings error at https://github.com/JuliaLang/julia/issues/27109? It happens on some systems and not others somewhat randomly.

**Comment by staticfloat** [comment]
I don't think so, because this is dying within `libgit2` initialization, whereas that issue is about the finalizers that are applied to certain libgit2 objects.

**Pull request merged**